src/Features/Core/Portable/Snippets/SnippetProviders/AbstractClassSnippetProvider.cs

```diff
@@ -0,0 +1,74 @@
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Snippets
+{
+    internal abstract class AbstractClassSnippetProvider : AbstractSnippetProvider
+    {
+        protected abstract void GetClassDeclaration(SyntaxNode node, out SyntaxToken identifier, out int cursorPosition);
+
+        public override string SnippetIdentifier => "class";
+
+        public override string SnippetDescription => FeaturesResources.class_;
+
+        protected override async Task<ImmutableArray<TextChange>> GenerateSnippetTextChangesAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var classDeclaration = generator.ClassDeclaration(NameGenerator.GenerateUniqueName("MyClass", name => semanticModel.LookupSymbols(position, name: name).IsEmpty));
+
+            var snippetTextChange = new TextChange(TextSpan.FromBounds(position, position), classDeclaration.NormalizeWhitespace().ToFullString());
+            return ImmutableArray.Create(snippetTextChange);
+        }
+
+        protected override Func<SyntaxNode?, bool> GetSnippetContainerFunction(ISyntaxFacts syntaxFacts)
+        {
+            return syntaxFacts.IsClassDeclaration;
+        }
+
+        protected override int GetTargetCaretPosition(ISyntaxFactsService syntaxFacts, SyntaxNode caretTarget)
+        {
+            GetClassDeclaration(caretTarget, out _, out var cursorPosition);
+            return cursorPosition;
+        }
+
+        protected override async Task<SyntaxNode> AnnotateNodesToReformatAsync(Document document,
+            SyntaxAnnotation findSnippetAnnotation, SyntaxAnnotation cursorAnnotation, int position, CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var snippetExpressionNode = FindAddedSnippetSyntaxNode(root, position, syntaxFacts.IsClassDeclaration);
+            if (snippetExpressionNode is null)
+            {
+                return root;
+            }
+
+            var reformatSnippetNode = snippetExpressionNode.WithAdditionalAnnotations(findSnippetAnnotation, cursorAnnotation, Simplifier.Annotation, Formatter.Annotation);
+            return root.ReplaceNode(snippetExpressionNode, reformatSnippetNode);
+        }
+
+        protected override ImmutableArray<SnippetPlaceholder> GetPlaceHolderLocationsList(SyntaxNode node, ISyntaxFacts syntaxFacts, CancellationToken cancellationToken)
+        {
+            using var _ = ArrayBuilder<SnippetPlaceholder>.GetInstance(out var arrayBuilder);
+            GetClassDeclaration(node, out var identifier, out var unusedValue);
```

Comment:
Nit:

```suggestion
            GetClassDeclaration(node, out var identifier, out _);
```

Reply:
unfortunately can't because of the line directly above using the discard :/